clang/test/Sema/builtin-counted-by-ref.c

```diff
@@ -0,0 +1,120 @@
+// RUN: %clang_cc1 -std=c99 -fsyntax-only -verify %s
+
+typedef unsigned long int size_t;
+
+int global_array[42];
+int global_int;
+
+struct fam_struct {
+  int x;
+  char count;
+  int array[] __attribute__((counted_by(count)));
+};
+
+void test1(struct fam_struct *ptr, int size, int idx) {
+  size_t size_of = sizeof(__builtin_counted_by_ref(ptr->array)); // ok
+
+  *__builtin_counted_by_ref(ptr->array) = size;             // ok
+  *__builtin_counted_by_ref(&ptr->array[idx]) = size;       // ok
+
+  {
+      size_t __ignored_assignment;
+      *_Generic(__builtin_counted_by_ref(ptr->array),
+               void *: &__ignored_assignment,
+               default: __builtin_counted_by_ref(ptr->array)) = 42; // ok
+  }
+}
+
+void test2(struct fam_struct *ptr, int idx) {
+  __builtin_counted_by_ref();                               // expected-error {{too few arguments to function call, expected 1, have 0}}
+  __builtin_counted_by_ref(ptr->array, ptr->x, ptr->count); // expected-error {{too many arguments to function call, expected 1, have 3}}
+}
+
+void test3(struct fam_struct *ptr, int idx) {
+  __builtin_counted_by_ref(ptr->x);                         // expected-error {{'__builtin_counted_by_ref' argument must reference a flexible array member}}
+  __builtin_counted_by_ref(&ptr->x);                        // expected-error {{'__builtin_counted_by_ref' argument must reference a flexible array member}}
+  __builtin_counted_by_ref(global_array);                   // expected-error {{'__builtin_counted_by_ref' argument must reference a flexible array member}}
+  __builtin_counted_by_ref(global_int);                     // expected-error {{'__builtin_counted_by_ref' argument must reference a flexible array member}}
+  __builtin_counted_by_ref(&global_int);                    // expected-error {{'__builtin_counted_by_ref' argument must reference a flexible array member}}
+}
+
+void test4(struct fam_struct *ptr, int idx) {
+  __builtin_counted_by_ref(&ptr->array[idx++]);             // expected-warning {{'__builtin_counted_by_ref' argument has side-effects that will be discarded}}
```

Comment:
discarded means the argument is not evaluated? Could you please add a codegen test for this warning case (since it's still supposed to build and generate some code).

Reply:
One already exists in tests/CodeGen/builtin-counted-by-ref.c `test4`.